src/platforms/common/px4_work_queue/WorkQueueManager.hpp

```diff
@@ -0,0 +1,100 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include <stdint.h>
+
+namespace px4
+{
+
+class WorkQueue; // forward declaration
+
+struct wq_config_t {
+	const char *name;
+	uint16_t stacksize;
+	int8_t relative_priority; // relative to max
+};
+
+namespace wq_configurations
+{
+static constexpr wq_config_t rate_ctrl{"wq:rate_ctrl", 1500, 0}; // PX4 inner loop highest priority
+
+static constexpr wq_config_t SPI1{"wq:SPI1", 1200, -1};
+static constexpr wq_config_t SPI2{"wq:SPI2", 1200, -2};
+static constexpr wq_config_t SPI3{"wq:SPI3", 1200, -3};
+static constexpr wq_config_t SPI4{"wq:SPI4", 1200, -4};
+static constexpr wq_config_t SPI5{"wq:SPI5", 1200, -5};
+static constexpr wq_config_t SPI6{"wq:SPI6", 1200, -6};
+
+static constexpr wq_config_t I2C1{"wq:I2C1", 1000, -7};
+static constexpr wq_config_t I2C2{"wq:I2C2", 1000, -8};
+static constexpr wq_config_t I2C3{"wq:I2C3", 1000, -9};
+static constexpr wq_config_t I2C4{"wq:I2C4", 1000, -10};
+
+static constexpr wq_config_t hp_default{"wq:hp_default", 1000, -11};
+static constexpr wq_config_t lp_default{"wq:lp_default", 1000, -50};
+
+static constexpr wq_config_t test1{"wq:test1", 800, 0};
+static constexpr wq_config_t test2{"wq:test2", 800, 0};
+
+} // namespace wq_configurations
+
+/**
+ * Start the work queue manager task.
+ */
+int WorkQueueManagerStart();
+
+/**
+ * Stop the work queue manager task.
+ */
+int WorkQueueManagerStop();
+
+/**
+ * Create (or find) a work queue with a particular configuration.
+ *
+ * @param new_wq		The work queue configuration (see WorkQueueManager.hpp).
+ * @return		A pointer to the WorkQueue, or nullptr on failure.
+ */
+WorkQueue *WorkQueueFindOrCreate(const wq_config_t &new_wq);
```

Comment:
What's the use case for this?

Reply:
The actual WorkQueues (threads) are only created as needed when a particular WorkItem is started.

Example from https://github.com/PX4/Firmware/pull/11571
 - mpu6000 start https://github.com/PX4/Firmware/pull/11571/commits/9ef32ef28c78c844c8595894431c9fc2562edf14#diff-65883d7b341346cb3c8e6763a8c8ec25R44
 - the desired work queue based on the interface (SPI1, etc) is passed to the ScheduledWorkItem constructor
 - https://github.com/PX4/Firmware/blob/e97dfec38b30ba0e21675abcae2316954d5b6695/src/platforms/common/px4_work_queue/WorkItem.cpp#L52-L65

Reply:
Ok got it.